src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -928,13 +921,17 @@ public ICollection MetadataNames
             {
                 get
                 {
-                    List<string> names = new List<string>((List<string>)CustomMetadataNames);
+                    CopyOnWritePropertyDictionary<ProjectMetadataInstance> metadataCollection = MetadataCollection;
+
+                    List<string> names = new List<string>(capacity: metadataCollection.Count + FileUtilities.ItemSpecModifiers.All.Length);
 
-                    foreach (string name in FileUtilities.ItemSpecModifiers.All)
+                    foreach (ProjectMetadataInstance metadatum in metadataCollection)
```

Comment:
```suggestion
                    names.AddRange(metadataCollection.Keys);
```

?

Reply:
(If the Keys method doesn't create a new object.)

Reply:
`Keys` is an `ICollection` so it must allocate linear storage. The underlying structure is an immutable dictionary, which is a tree. Enumerating it will be more efficient.